src/main/drivers/system.c

```diff
@@ -60,16 +60,60 @@ void cycleCounterInit(void)
     usTicks = clocks.SYSCLK_Frequency / 1000000;
 }
 
+#include "common/atomic.h"
+#include "debug.h"
+
+static volatile int sysTickRollover = 0;
+
 // SysTick
 void SysTick_Handler(void)
 {
-    sysTickUptime++;
+    ATOMIC_BLOCK(NVIC_PRIO_MAX) {
+        sysTickUptime++;
+        sysTickRollover = 0;
+        *(volatile uint32_t *)&(SysTick->CTRL);
+    }
+}
+
+uint32_t microsISR(void)
+{
+    register uint32_t ms, cycle_cnt;
+
+    ATOMIC_BLOCK(NVIC_PRIO_MAX) {
+        cycle_cnt = SysTick->VAL;
+
+        if (SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk) {
+            // Update pending.
+            // Remember it for calls within the same rollover period
+            // (Will be cleared when serviced).
+
+            // XXX Do we care for multiple rollovers?
+            // sysTickUptime would be incorrect in that case...
+
+            ++sysTickRollover;
+
+            // Read VAL again to ensure the value is read after the rollover.
+
+            cycle_cnt = SysTick->VAL;
+        }
+
+        ms = sysTickUptime;
+    }
+
+    return ((ms + sysTickRollover) * 1000) + (usTicks * 1000 - cycle_cnt) / usTicks;
 }
 
 // Return system uptime in microseconds (rollover in 70minutes)
 uint32_t micros(void)
 {
     register uint32_t ms, cycle_cnt;
+
+    // Call microsISR() in interrupt context
+
+    if (SCB->ICSR & SCB_ICSR_VECTACTIVE_Msk) {
```

Comment:
`microsISR` is probably fine. 
As implemented, this function can be called in any context and will return correct value, so it is pretty safe. Only disadvantage is that it is disabling interrupts/is a bit slower